### PR TITLE
[doc] Add user utilities doc

### LINKED ISSUE #1269: [Doc] add more details for debug utils and developer utils
**Concisely describe the proposed feature**
Sometimes I felt a little hard to develop Taichi because Our "developer utils" is not as powerful as we imagine.

For example `ti.info` this function lacks a clear statement, which means a fresh developer cannot learn from the doc for utils he may need. 

And I realize that there are lots of functions that haven't been documented well, for example, `ti.ti_print`.

**A clear and concise description of what you want.**

I hope to consummate the "developer utils" parts such as for `ti.info`, add more details in it so that new developer can feel easier

I don't know whether `ti.ti_print()` has been deprecated or not. If not, I can add some details to it.

Also, I have realized that the backend of `ti.info` is [spdlog](https://github.com/gabime/spdlog). I am using it as well in my own project. spdlog should also support logging to a `.log` file, do we need to come up with a way to enable this feature in Taichi python-binding? 

If we want to use Taichi in a big project, logging to file should be essential. Btw, if this has been supported, I can add a doc for it as well.

***
2020/06/18 update:

So after a brief discussion with @archibate , I realize that:

1. `ti_ti_print` is not a developer util but debug util. 
2. Maybe we can try to add a new section for debug util, including but not limited to `ti_print`, `ti_static_print` and others about `profiler`
3. I hope to add more details for `ti.info` in Developer util.

HDYT?

### PR DESCRIPTION
**Pull request by Rullec**
Related issue = Close #1269 

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

This draft PR is for user utilities. I simply add a `user_utilities` in `MISCELLANEOUS` part. Current Changes shows below.

1. More Explanations on ~~`ti.info`,~~ this logging module
2. Document our `Profiler` module for the first time
3. ~~say something about `assertation`~~: finished in #1344 

I still need to do:
## 1. Logging
1. ~~I am not satisfied with `ti.info` this part. I think it can do more, for example display a tensor's value (under the help of precessor AST)~~
2. ~~A potential bug in `ti.error` or `ti.critical`, maybe we need to add a test script for this functionality.~~
3. ~~`ti.critical` and `ti.CRITICAL` are missing, why?~~ 
4. ~~``ti.info`` is only executed once in compile-time. So what are the suitable scenarios for this command?: WIP, the same reason as ``ti.static_print`` in #1344~~
5. ~~``ti.static_print`` is executed in compile-time as well. What are the suitable scenarios?~~ : finished in #1344 
6. ~~I think we need to say that print doesn't support f-string, otherwise lots of guys need to try it and get confused. the same as ``ti.info`` and ``ti.static_print``: maybe finished in #1344~~ 

## 2. Profiling 
1. ~~Our profiler is a little weak relatively. It can not do some very basic profiling~~
2. ~~Profiler doesn't work in `ti.cpu`: Need to test.~~
3. ~~Two profilers are needed to be documented.~~
4. ~~To distinguish our 2 profilers `ScopedProfiler` and `ProfilerBase` #1261, I used these 2 names directly in this doc. But they are hidden deeply in Taichi's core and should not be exposed to the end-user. Is there a better way to do that?~~
5. We need to explain the output of `profilers`, explain what these statistics stand for in detail. (Sadly I cannot understand all of them as well.)

## 3. Asserting
1. ~~doc python built-in `assert`~~
2. ~~implement `ti.static_assert`. What feature do we want exactly in `ti.static_assert`, I am a little confused... :( Is it an assertation that works in runtime as well??~~: finished in #1344 
3. ~~doc `ti.static_assert`.~~ : finished in #1344 

All of these conculsions are my own and only temporary. Some of them may not be true but I need to verify it again and dig deeper.

### CODE REVIEW

File: docs/user_utilities.rst [hunk 1]
```diff
@@ -0,0 +1,109 @@
+User utilities
```
Comment (archibate) [inline docs/user_utilities.rst:1]: Suggestion: Debugging utilities

File: docs/user_utilities.rst [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    import taichi as ti
+    ti.init(arch = ti.cpu)
+
+    ti.set_logging_level(ti.INFO)
+    var = ti.var(ti.f32, shape = 1)
+
+    @ti.kernel
+    def compute():
+        var[0] = 1.0
+        ti.info(f"set var[0] = 1.0")
```
Comment (archibate) [inline docs/user_utilities.rst:46]: Let's be clear that this is only called at compile-time. i.e. If you recall compute, it won't reprint the string. (same as ti.static_print)

File: docs/user_utilities.rst [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    ti.cfg.kernel_profiler = True
+    compute()
+    ti.kernel_profiler_print()
+    ti.kernel_profiler_clear()
+
+TODO: 
+
+    1.These explanations are not clear enough. A better example is essential.
+
+    2. Can out profiler do more things?
```
Comment (archibate) [inline docs/user_utilities.rst:99]: Yes, we have ti.profiler_start for general purpose profilijg, please doc that too.

Comment (archibate) [inline docs/user_utilities.rst:99]: Yes, we have ti.profiler_start for general purpose profilijg, please doc that too.

File: docs/user_utilities.rst [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+Assertation
+-----------
+Assertation can make your Taichi program more reliable and help you debug easily.
+
+TODO:
+
+    1. add ``ti.static_assert``
```
Comment (archibate) [inline docs/user_utilities.rst:107]: Also add non-static `assert` doc.

File: docs/user_utilities.rst [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+.. Note::
+
+    The usages of ``ti.trace``, ``ti.debug``, ``ti.error``, ``ti.critical`` are almost similar as ``ti.info``.
+
+TODO: 
+
+    1. more explanations and examples in this logging section. can we log a taichi tensor here, why and how?
+
+    2. ``ti.error`` and ``ti.critical`` tested fail. I need to check it again
```
Comment (archibate) [inline docs/user_utilities.rst:58]: In fact, ti.error will purposely terminate the program, IIRC.

Comment (Rullec) [reply docs/user_utilities.rst:58]: Yes, I think you are correct according to my experiment. `ti.error` will cause a crash.

I can fully understand this design, but not very agree with it. Our loggging utility should **focus on logging** but not try to control the workflow of ANY `Taichi` program. 

For example, If I say:
```
ti.error("here is an error!")
```
I simply want to **show this error** to the screen but not crash in this place. It's also spdlog recommending.

HDYT?

File: docs/user_utilities.rst (hunk 3 above)
Comment (Rullec) [reply docs/user_utilities.rst:99]: Are you referring to the `LLVMRuntime_set_profiler_start`?

File: docs/user_utilities.rst (hunk 2 above)
Comment (Rullec) [reply docs/user_utilities.rst:46]: Agree. 

I have an observation, would you please help me to check whether my understanding is correct?

According to the `Debugging` part in this [doc](https://taichi.readthedocs.io/en/stable/syntax.html), It seems that all of the `print` will be transfered to `ti.static_print` by the AST preprocessor, so that they will be executed in run-time. That's good.

But it doesn't include logging utils such `ti.info`, these `ti.info` are only called at compiled-time but not runtime.
***
So we cannot do any logging `ti.info` in kernel runtime, right? :( only print is supported in runtime is a little trivial, and IMO the debugging experince can be nicer if we support `ti.info` in runtime.

Basically, why wasn't `ti.info` be supported in runtime? It looks not very hard. Would you be happy to introduce the idea behind this design? :D

File: docs/user_utilities.rst (hunk 1 above)
Comment (Rullec) [reply docs/user_utilities.rst:1]: Good suggestion

File: docs/user_utilities.rst (hunk 5 above)
Comment (archibate) [reply docs/user_utilities.rst:58]: That could be hard, since there's already many C++ code assuming `TI_ERROR` to be program-terminating. IMHO if user want to show error without terminate, they may use `ti.warn` or `ti.critical` instead.

File: docs/user_utilities.rst (hunk 3 above)
Comment (archibate) [reply docs/user_utilities.rst:99]: Oh sorry, didn't notice `ti.profiler_start` is remove in [#1261](https://github.com/taichi-dev/taichi/pull/1261/files#diff-637de6237b50874e58c91560edb063e4L38-L41)

File: docs/user_utilities.rst (hunk 2 above)
Comment (archibate) [reply docs/user_utilities.rst:46]: > It looks not very hard

No, **copying strings from GPU to CPU** is harder than you may think. Ancient GPUs doesn't even support integer, not to mention to implement runtime f-string...

In fact, the only difference between `ti.info` and `print` is its color and f-string.. not sure why `ti.info` superior.

File: docs/user_utilities.rst (hunk 3 above)
Comment (Rullec) [reply docs/user_utilities.rst:99]: Thanks for your nice guide!
I also notice in #1261 that we have 2 profiling system as yuanming said,
1. ProfilerBase for kernels
2. ScopedProfiler for some host tasks?

Are they all developed for user profiling? If yes, I can documated them both.

File: docs/user_utilities.rst (hunk 5 above)
Comment (Rullec) [reply docs/user_utilities.rst:58]: Great, so let's keep `ti.error` terminte the program and I will documted it on stress emphasize. :D

File: docs/user_utilities.rst (hunk 2 above)
Comment (Rullec) [reply docs/user_utilities.rst:46]: Thanks for your clear explanation! I get the point. Yes, I fully understand your meaning. ;-) 

The difference is just the timestamp(color is relatively unimportant). Let's assume we want to do a simulation which can cost for a whole day (It's quite common in mechanical and chemstry simulation, most of them are in reseach cases), If we want to find the bug after waking up in the morning, a timestamp or screenshot can be helpful IMAO.

Comment (archibate) [reply docs/user_utilities.rst:46]: I think they could do `ti.info` **between kernel launches**.
In fact, `print` in kernel also won't print until kernel ends as my explanation goes above.
So there's no difference for this usage.

If you want a timestamp in `print`, I think `print(ti.timestamp(), 'x =', x)` is a good approach (TBD).

Comment (Rullec) [reply docs/user_utilities.rst:46]: Thanks! 

I will invesigate this case further more. I need to get a global view of Taichi in order to understand it, instead of floating on the surface. XD

**Comment by yuanming-hu** [comment]
Sorry about my delayed comments.

> * Our profiler is a little weak relatively. It can not do some very basic profiling
> * Profiler doesn't work in `ti.cpu`

We have two sets of profilers: #1261 

> * implement `ti.static_assert`

This will be very helpful!

> * `ti.critical` and `ti.CRITICAL` are missing, why?

I guess that's because we rarely use them. For now, `ti.error` seems to be the highest level of erroneous logging. Adding `critical` would be helpful. Thank you!

**Comment by Rullec** [comment]
Thanks a lot for your powerful guide! So that I can investigate these 2 profilers deeply and documente them!

Agree, I also think `ti.ERROR` is enough for us because it even can crash the program :O

Btw, #1312 simply fix some typos in `export_results.rst` and the changeset is quite small. Would you be happy to check that in your spare time?

**Comment by Rullec** [comment]
> We have two sets of profilers: #1261

The design of `ScopedProfiler` is so exquisite! Very enjoyable to read it.

But I can only see its work in a single thread. Would you please give me some guides on a multi-threads case?

**Comment by Rullec** [comment]
Update:

First sorry: I do ``git merge --rebase master`` in my local repo in order to catch up with the latest change, but now the commit history is messed out. Do I need to fix it? If yes, what can I do? :(

Here is a new version ``user_utilities.rst``! Though it is far less perfect, but it covers more than the latter version. 

I met with some new questions and got stuck when writing this doc. @archibate  Would you be happy to give me some guidence? The questions are maintained in the beginning part. Thanks in advance!

**Comment by archibate** [comment]
On my way!

**Comment by Rullec** [comment]
> On my way!
> EDIT: a little tricky, will first go fix this tmr, good night!

#1344  is very amazing! So this issue #1295  can simply focus on the other 2 parts. :-)

**Comment by archibate** [comment]
@Rullec Fixed now, run ``git pull --force`` to go ahead.

**Comment by Rullec** [comment]
Thanks!

彭于斌 <notifications@github.com> 于2020年6月28日周日 上午11:03写道：

**Comment by Rullec** [comment]
Update: Now there is a new version. Would be please give me some suggestions?

**Comment by Rullec** [comment]
Hi all,
  Is there anything I can improve now? Suggestions are welcome!

File: docs/user_utilities.rst [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+    In spite of this, the output of ``ti.error()`` is still ``stdout`` but not ``stderr``.
+
+Here is an example:
+
+.. code-block:: python
+
+    import taichi as ti
+
+    ti.init()
+    ti.set_logging_level(ti.ERROR)
```
Comment (archibate) [inline docs/user_utilities.rst:88]: Let's remove this or user would think loglev has to be error to make error work.

File: docs/user_utilities.rst [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+
+    @ti.kernel
+    def compute():
+        var[0] = 1.0
+        print(f" set var[0] = {var[0]}")
+
+    compute()
+    ti.core.print_profile_info()
+
+``ti.core.print_profile_info()`` will output statistics in a hierarchical format, with different colors for different levels.
```
Comment (archibate) [inline docs/user_utilities.rst:120]: OFT: Can we make this ti.print_profile_info?

File: docs/user_utilities.rst [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+
+    compute()
+    ti.kernel_profiler_print()
+
+Its output is like:
+
+::
+
+    [ 22.73%] jit_evaluator_0_kernel_0_serial             min   0.001 ms   avg   0.001 ms   max   0.001 ms   total   0.000 s [      1x]
+    [  0.00%] jit_evaluator_1_kernel_1_serial             min   0.000 ms   avg   0.000 ms   max   0.000 ms   total   0.000 s [      1x]
```
Comment (archibate) [inline docs/user_utilities.rst:153]: OFT: Can we make jit&accessor kernels invisible by default? They are usually 0 and confusing to end-users.

File: docs/user_utilities.rst [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+############
+
+1. ``ProfilerBase`` records the cost of kernel launches on devices. ``kernel_profiler`` must be set to ``True`` in ``ti.init`` to enable this profiler.
+
+2. When all the computation is finished in your program, call ``ti.kernel_profiler_print()`` to show to profiling result. For example:
+
+.. code-block :: python
+    :emphasize-lines: 1, 9
+
+    ti.init(ti.cpu, kernel_profiler = True)
```
Comment (archibate) [inline docs/user_utilities.rst:138]: Please leave no space around `=`, see yapf.

**Comment by archibate** [comment]
Hello? @Rullec are you still there?

**Comment by Rullec** [comment]
Sorry for my late reply, I'm on ddl for these 2 weeks! I will update it
hopefully 2 hrs later.

彭于斌 <notifications@github.com> 于2020年7月5日周日 下午12:52写道：

File: docs/user_utilities.rst (hunk 6 above)
Comment (Rullec) [reply docs/user_utilities.rst:88]: Great!

File: docs/user_utilities.rst (hunk 7 above)
Comment (Rullec) [reply docs/user_utilities.rst:120]: Done, update comming soon.

File: docs/user_utilities.rst (hunk 9 above)
Comment (Rullec) [reply docs/user_utilities.rst:148]: Thanks, a formatter has been introduced in now.

**Comment by Rullec** [comment]
Thanks @archibate  

All suggestions has been adopted & tested in this new commit, but I feel hard to know how to set jit & aceesor output closed by default in `ScopedProfiler`. 

Would you please give me some advice?

**Comment by archibate** [comment]
> All suggestions has been adopted & tested in this new commit, but I feel hard to know how to set jit & aceesor output closed by default in ScopedProfiler.

Oh I see, that would be hard, let's try do it in a separate PR, and focus on doc in this PR.

File: docs/user_utilities.rst [hunk 10]
```diff
@@ -34,16 +34,18 @@ Based on spdlog, Taichi provides a series of APIs for logging. Note that these f
...
 
     ti.set_logging_level(ti.INFO)
-    var = ti.var(ti.f32, shape = 1)
+    var = ti.var(ti.f32, shape=1)
+
 
     @ti.kernel
     def compute():
         var[0] = 1.0
         ti.info(f"set var[0] = 1.0")
```
Comment (archibate) [inline docs/user_utilities.rst:46]:
```suggestion
```
Let's be clear that `ti.info` cannot be used in Python-scope.
They should be moved to **developer utilities**, the logging system is mainly used by Taichi developers, not end-users.

Comment (Rullec) [reply docs/user_utilities.rst:46]: Nice, WIP

Comment (Rullec) [reply docs/user_utilities.rst:46]: > Let's be clear that `ti.info` cannot be used in Python-scope.

In this example, it seems `ti.info` is in Taichi-scope... :O?

**Comment by Rullec** [comment]
Update:
1. Move all contents to `utilities.rst`
2. fix some typos

File: docs/utilities.rst [hunk 11]
```diff
@@ -1,25 +1,168 @@
...
+
+Profiler
+--------
+
+Taichi profiler can help you count the time cost of your program. There are two profilers in Taichi available at this moment: ``ScopedProfiler`` and ``ProfilerBase``.
+
+ScopedProfiler
+##############
+
+1. ``ScopedProfiler`` evaluates the time spent for each step of the **host tasks** during execution hierarchically.
```
Comment (yuanming-hu) [inline docs/utilities.rst:109]:
```suggestion
1. ``ScopedProfiler`` measures time spent on the **host tasks** hierarchically.
```

File: docs/utilities.rst [hunk 12]
```diff
@@ -1,25 +1,168 @@
...
+
+2. Moreover, this profiler is turned on automatically and no extra work is needed except writing ``ti.print_profile_info()`` at the end of your Taichi program. For example:
```
Comment (yuanming-hu) [inline docs/utilities.rst:111]:
```suggestion
2. This profiler is turned on automatically. To show its results, call ``ti.print_profile_info()``. For example:
```

File: docs/utilities.rst [hunk 13]
```diff
@@ -1,25 +1,168 @@
...
+    import taichi as ti
+
+    ti.init(arch=ti.cpu)
+    var = ti.var(ti.f32, shape=1)
+
+
+    @ti.kernel
+    def compute():
+        var[0] = 1.0
+        print(f" set var[0] = {var[0]}")
```
Comment (yuanming-hu) [inline docs/utilities.rst:124]:
```suggestion
        print(f"set var[0] =", var[0])
```

File: docs/utilities.rst [hunk 14]
```diff
@@ -1,25 +1,168 @@
...
+
+
+    compute()
+    ti.print_profile_info()
+
+
+``ti.print_profile_info()`` will output statistics in a hierarchical format, with different colors for different levels.
```
Comment (yuanming-hu) [inline docs/utilities.rst:131]:
```suggestion
``ti.print_profile_info()`` prints profiling results in a hierarchical format.
```

File: docs/utilities.rst [hunk 15]
```diff
@@ -1,25 +1,168 @@
...
+
+.. Note::
+
+    ``ScopedProfiler`` is a c++ class inside the core of Taichi, so it is invisible in Python-scope.We mention its name here is just to make a distinction with another profiler.
```
Comment (yuanming-hu) [inline docs/utilities.rst:135]:
```suggestion
    ``ScopedProfiler`` is a C++ class in the core of Taichi. It is not exposed to Python users.
```

File: docs/utilities.rst [hunk 16]
```diff
@@ -1,25 +1,168 @@
...
+
+ProfilerBase
```
Comment (yuanming-hu) [inline docs/utilities.rst:137]: This should be called `KernelProfiler`.

File: docs/utilities.rst [hunk 17]
```diff
@@ -1,25 +1,168 @@
...
+############
+
+1. ``ProfilerBase`` records the cost of kernel launches on devices. ``kernel_profiler`` must be set to ``True`` in ``ti.init`` to enable this profiler.
```
Comment (yuanming-hu) [inline docs/utilities.rst:140]:
```suggestion
1. ``KernelProfiler`` records the costs of Taichi kernels on devices. To enable this profiler, set ``kernel_profiler=True`` in ``ti.init``.
```

File: docs/utilities.rst [hunk 18]
```diff
@@ -1,25 +1,168 @@
...
+
+2. When all the computations are finished in your program, call ``ti.kernel_profiler_print()`` to show the profiling result. For example:
```
Comment (yuanming-hu) [inline docs/utilities.rst:142]:
```suggestion
2. Call ``ti.kernel_profiler_print()`` to show the kernel profiling result. For example:
```

No need to wait for all computations are finished.

File: docs/utilities.rst [hunk 19]
```diff
@@ -1,25 +1,168 @@
...
+
 .. code-block:: python
+    :emphasize-lines: 3, 11
+
+    import taichi as ti
+
+    ti.init(ti.cpu, kernel_profiler = True)
```
Comment (yuanming-hu) [inline docs/utilities.rst:149]:
```suggestion
    ti.init(ti.cpu, kernel_profiler=True)
```

File: docs/utilities.rst [hunk 20]
```diff
@@ -1,25 +1,168 @@
...
+    var = ti.var(ti.f32, shape = 1)
```
Comment (yuanming-hu) [inline docs/utilities.rst:150]:
```suggestion
    var = ti.var(ti.f32, shape=1)
```

File: docs/utilities.rst [hunk 21]
```diff
@@ -1,25 +1,168 @@
...
+
+    @ti.kernel
+    def compute():
+        var[0] = 1.0
+
+    compute()
+    ti.kernel_profiler_print()
+
+Its output is like:
```
Comment (yuanming-hu) [inline docs/utilities.rst:159]:
```suggestion
The outputs would be:
```

**Comment by Rullec** [comment]
> Cool! LGTM in general. I took a pass and fixed a few wording issues. Please take your time and no rush at all.
> 
> > Hello? @Rullec are you still there?
> 
> @archibate please be courteous and friendly. We are an open-source community and nobody is obligated to contribute. Keep in mind that most people have their daily work to do. Thanks! :-)

Thanks both @archibate and @yuanming-hu , for your kindly comprehension and all of these powerful modifications. :D

I will consider how to revise this doc again, and hopefully a new version can come soon.

File: docs/utilities.rst [hunk 22]
```diff
@@ -1,25 +1,168 @@
...
+    Taichi logging APIs only support standard output now.
+
+.. function:: ti.set_logging_level(level)
+
+    :parameter level: (string) a valid logging level
+
+    - This function is used to set the level of logging. Currently the log levels in Taichi are sorted as ``ti.TRACE``, ``ti.DEBUG``, ``ti.INFO``, ``ti.WARN`` and ``ti.ERROR`` in descending order. The default logging level is ``ti.INFO``.
+
+    - These logging levels are essentially strings ``trace``, ``debug``, ``info``, ``warn`` and ``error`` respectively.
+
```
[the remark this answers is not in the document]
Comment (Rullec) [reply docs/utilities.rst:25]: Okay :)

File: docs/utilities.rst [hunk 23]
```diff
@@ -1,25 +1,168 @@
...
+.. function:: ti.error(info)
+
+    :parameter info: (string) logging info
+
+    This function prints the input string in red color to stdout in any logging level and **only in Taichi-scope**.
+
+.. warning::
+
+    Note that ``ti.error`` will crash your program and throws an exception ``RuntimeError``.
```
[the remark this answers is not in the document]
Comment (Rullec) [reply docs/utilities.rst:84]: Good idea! Please let me simplify this part and submit a new version.

File: docs/utilities.rst (hunk 16 above)
Comment (Rullec) [reply docs/utilities.rst:137]: Oh! The old `ProfilerBase` has been renamed to `KernelProfilerBase`, I didn't manage to notice it...

File: docs/utilities.rst (hunk 20 above)
Comment (Rullec) [reply docs/utilities.rst:150]: I will format the example code even in the doc next time 👍 :(

**Comment by Rullec** [comment]
Hi all, here is a new version. Most of problems we mentioned above has been treated in it.

But I still have a question, @yuanming-hu said that **`ti.info` should not be used in `ti.kernel`**, would you please tell me the story behind it? Because it doesn't throw an exception and also works the same as my expection. Thanks!

**Comment by archibate** [comment]
> https://github.com/taichi-dev/taichi/pull/1295#issuecomment-655637497

Just invoke the kernel **twice**, and you'll find it's only printed **once**.

**Comment by Rullec** [comment]
> Just invoke the kernel **twice**, and you'll find it's only printed **once**.

But isn't that what we want? For example,

```
import taichi as ti
ti.init(arch=ti.cpu)

ti.set_logging_level(ti.INFO)
mat = ti.var(dt=ti.f32, shape=(5, 5))


@ti.func
def calc(i: ti.int32, j: ti.int32):
    ti.info(f"set var in ti.func")
    mat[i, j] = i * j


@ti.kernel
def compute():
    calc(0, 0)
    calc(1, 1)
    calc(2, 2)
    ti.info(f"set var in ti.kernel")


compute()
```

output:
```
[I 07/09/20 01:29:58.039] [main.py:calc@11] set var in ti.func
[I 07/09/20 01:29:58.040] [main.py:calc@11] set var in ti.func
[I 07/09/20 01:29:58.040] [main.py:calc@11] set var in ti.func
[I 07/09/20 01:29:58.040] [main.py:calc@11] set var in ti.func
[I 07/09/20 01:29:58.041] [main.py:compute@22] set var in ti.kernel
```

It has been pointed out in our doc `These functions should only be called at compile-time instead of run-time`.
which indicates that `ti.info` inside of `ti.kernel` only runs for a single time when kernels are invoked, aka `compile-time`. 

So I feel a little hard to understand why @yuanming-hu  said using `ti.info` in `ti.kernel` is misleading :).

**Comment by archibate** [comment]
```py
import taichi as ti
ti.init(arch=ti.cpu)

ti.set_logging_level(ti.INFO)
mat = ti.var(dt=ti.f32, shape=(5, 5))


@ti.func
def calc(i: ti.int32, j: ti.int32):
    ti.info(f"set var in ti.func")
    mat[i, j] = i * j


@ti.kernel
def compute():
    calc(0, 0)
    calc(1, 1)
    calc(2, 2)
    ti.info(f"set var in ti.kernel")


compute()
compute()
compute()
compute()
compute()
compute()
compute()
compute()
```
output:
```
[Taichi] mode=development
[Taichi] preparing sandbox at /tmp/taichi-cudg5ylh
[Taichi] <dev mode>, llvm 10.0.0, commit 8dbcc3b8, python 3.8.3
[Taichi] Starting on arch=x64
[I 07/09/20 11:52:15.340] [a.py:calc@10] set var in ti.func
[I 07/09/20 11:52:15.341] [a.py:calc@10] set var in ti.func
[I 07/09/20 11:52:15.342] [a.py:calc@10] set var in ti.func
[I 07/09/20 11:52:15.342] [a.py:compute@19] set var in ti.kernel
```

**Comment by Rullec** [comment]
Yes, It seems we reaches an agreement. Here is a new example, would it be better?

File: docs/utilities.rst [hunk 24]
```diff
@@ -1,25 +1,185 @@
...
+    ti.init()
+    ti.set_logging_level(ti.INFO)
+
+    try:
+        ti.error("Fatal error. Exiting now...")
+    except RuntimeError as err:
+        print(err)
+
+
+Profiler
```
Comment (archibate) [inline docs/utilities.rst:115]: Profiler should be shared by devs and end-users, maybe we can have a separate .rst for it?

File: docs/utilities.rst [hunk 25]
```diff
@@ -1,25 +1,185 @@
 Developer utilities
 ===================
 
+
+This section provides a detailed description of some commonly used utilities for Taichi developers.
+
 Logging
```
Comment (archibate) [inline docs/utilities.rst:7]: Hi, sorry about dismissing your changes, I have already done with `Logging` in #1475 :)

File: docs/utilities.rst (hunk 24 above)
Comment (Rullec) [reply docs/utilities.rst:115]: IMO I think it's enough in `utilities.rst` at this moment.

File: docs/utilities.rst (hunk 25 above)
Comment (Rullec) [reply docs/utilities.rst:7]: May you merge these 2 changes into a single one? I also have no idea about how to deal with this conflict.

Comment (archibate) [reply docs/utilities.rst:7]: To prevent conflicts with #1475, you may move the `Profiler` to another section like `profiler.rst` :) I'll try to merge conflicts in `Logging` section once one of these two PR is merged.

**Comment by Rullec** [comment]
Finished, as @archibate adviced, I remove all contents about `logging` :(. It is not a good feeling...I hope next time maybe I can avoid this duplicate work

File: docs/profiler.rst [hunk 26]
```diff
@@ -0,0 +1,69 @@
...
+    import taichi as ti
+
+    ti.init(arch=ti.cpu)
+    var = ti.var(ti.f32, shape=1)
+
+
+    @ti.kernel
+    def compute():
+        var[0] = 1.0
+        print(f"set var[0] =", var[0])
```
Comment (yuanming-hu) [inline docs/profiler.rst:25]:
```suggestion
        print("Setting var[0] =", var[0])
```

File: docs/profiler.rst [hunk 27]
```diff
@@ -0,0 +1,69 @@
...
+========
+
+Taichi's profiler can help you analyze the run-time cost of your program. There are two profiling systems in Taichi: ``ScopedProfiler`` and ``KernelProfiler``.
+
+ScopedProfiler
+##############
+
+1. ``ScopedProfiler`` measures time spent on the **host tasks** hierarchically.
+
+2. This profiler is turned on automatically. To show its results, call ``ti.print_profile_info()``. For example:
```
Comment (yuanming-hu) [inline docs/profiler.rst:12]:
```suggestion
2. This profiler is automatically on. To show its results, call ``ti.print_profile_info()``. For example:
```

**Comment by Rullec** [comment]
Thank you all for your work as well!
